### ISSUE TITLE
Upgrading to TypeScript 3.7 yields vue-router type error with composition-api

### ISSUE THREAD

**Issue by aaf-ww** [vuejs/vue-router#3019]
### Version
3.1.3

### Reproduction link
[https://github.com/aaf-ww/vue-router-typescript-bug](https://github.com/aaf-ww/vue-router-typescript-bug)

### Steps to reproduce
Scaffold a new project with CLI (TypeScript/Babel/vue-router/etc), install @vue/composition-api then upgrade to TypeScript 3.7.2
```
vue create project-name (select all options)
npm install @vue/composition-api
npm update typescript@latest -D
```

### What is expected?
No type errors.

### What is actually happening?
ERROR in C:/Users/x/Desktop/source/vue-router-typescript-bug/src/router/index.ts(24,30):
24:30 Argument of type '{ mode: "history"; base: any; routes: ({ path: string; name: string; component: VueProxy<ComponentPropsOptions<Data>, Data>; } | { path: string; name: string; component: () => Promise<...>; })[]; }' is not assignable to parameter of type 'RouterOptions'.
  Types of property 'routes' are incompatible.
    Type '({ path: string; name: string; component: VueProxy<ComponentPropsOptions<Data>, Data>; } | { path: string; name: string; component: () => Promise<typeof import("C:/Users/x/Desktop/source/vue-router-typescript-bug/src/views/About.vue")>; })[]' is not assignable to type 'RouteConfig[]'.
      Type '{ path: string; name: string; component: VueProxy<ComponentPropsOptions<Data>, Data>; } | { path: string; name: string; component: () => Promise<typeof import("C:/Users/x/Desktop/source/vue-router-typescript-bug/src/views/About.vue")>; }' is not assignable to type 'RouteConfig'.
        Type '{ path: string; name: string; component: VueProxy<ComponentPropsOptions<Data>, Data>; }' is not assignable to type 'RouteConfig'.
          Types of property 'component' are incompatible.
            Type 'VueProxy<ComponentPropsOptions<Data>, Data>' is not assignable to type 'VueConstructor<Vue> | ComponentOptions<Vue, DefaultData<Vue>, DefaultMethods<Vue>, DefaultComputed, PropsDefinition<...>, Record<...>> | AsyncComponentPromise<...> | AsyncComponentFactory<...> | undefined'.
              Type 'ComponentOptions<Vue, { [x: string]: unknown; }, never, never, ComponentPropsOptions<Data>, ExtractPropTypes<ComponentPropsOptions<Data>, false>> & VueConstructorProxy<...>' is missing the following properties from type 'VueConstructor<Vue>': extend, nextTick, set, delete, and 9 more.
    22 | ];
    23 |
  > 24 | const router = new VueRouter({
       |                              ^
    25 |   mode: "history",
    26 |   base: process.env.BASE_URL,
    27 |   routes
Version: typescript 3.7.2

**Comment by Domenico-Pirozzi** [comment]
I have the same issue. Any news?

**Comment by s-hashimoto00** [comment]
This issue looks the same as https://github.com/vuejs/composition-api/issues/185

The same problem occurred, but it was solved by defining an empty “props”.
```ts
export default createComponent({
  props: {}
})
```

**Comment by lmiller1990** [comment]
An alternative until this is fixed is:

```
// @ts-ignore <- ignore it
const router = new VueRouter({
  routes,
  mode: 'history'
})
```

This may seem hacky, but I think it's better than adding empty `props` key to every component (did not work for me).

**Comment by keithn** [comment]
> This may seem hacky, but I think it's better than adding empty `props` key to every component (did not work for me).

@lmiller1990  did you try the suggestion from the post above yours to add ```props:{}``` ?  seemed to work for me

**Comment by lmiller1990** [comment]
I did, but then I need to add an empty `props` key to a large number of components. I'm happy to just `ts-ignore` for now. Wonder what the fix is for this? Might take a look and try and make a PR.

**Comment by LinusBorg** [comment]
typecasting the imported component also works as a workaround:
```typescript
{
  path: '/users',
  component: Users as VueConstructor<Vue>
}
```

**Comment by kouchanne** [comment]
I worked around this problem by this code.

```typescript
{
  path: '/users',
  component: () =>
      import("../views/User.vue")
}
```

**Comment by kristinruben** [comment]
@posva I noticed you added the `contribution welcome` tag! I am interested in working on TypeScript/Vue 3 compatibility - will have time to get started in the next week.

**Comment by LinusBorg** [comment]
Hey Kristin, awesome! Just to make sure: this issue is about compatibility with the Vue 2 plugin, not Vue 3.

But you're welcome to contribute to both :)

**Comment by posva** [comment]
It may be that the problem is at https://github.com/vuejs/composition-api/issues/185 too, ending up in no contribution here

**Comment by haoqunjiang** [comment]
I believe it's fixed in `@vue/composition-api` v0.4

**Comment by DarkLite1** [comment]
I 'm new to TypeScript and came here because think this is related. 
```javascript
export default function ({ store }) {
  const Router = new VueRouter({
    scrollBehavior: () => ({ x: 0, y: 0 }),
    routes,
 
    // Leave these as they are and change in quasar.conf.js instead!
    // quasar.conf.js -> build -> vueRouterMode
    // quasar.conf.js -> build -> publicPath
    mode: process.env.VUE_ROUTER_MODE,
    base: process.env.VUE_ROUTER_BASE
  })
```
The `mode` is underlined saying
![image](https://user-images.githubusercontent.com/10828570/80955972-8f16c600-8e00-11ea-9a64-c3a1856341d9.png)

What is the correct way to avoid this error?
